<div style="border-left:5px solid #de4d43; border-radius:2pt; color:#037691; padding:16px; background-color: #a7ced6">
<div style="color:#de4d43;">
    
# Consulter et mettre à jour une base de données
</div>
    
#### d'après bac 2022 Centres étrangers, J1, Ex. 4
<span class="badge" style="color:#ffffff; background-color: #037691"> sql</span> <span class="badge" style="color:#ffffff; background-color: #037691"> base de données</span>
</div>

Dans le cadre d'une étude sur le réchauffement climatique, un centre météorologique rassemble des données. On considère que la base de données contient deux relations :
- La relation **`Centres`** qui contient l'identifiant des centres météorologiques, la ville, la latitude, la longitude et l'altitude du centre. 
- La relation **`Mesures`** qui contient l'identifiant de la mesure, l'identifiant du centre, la date de la mesure, la température, la pression et la pluviométrie mesurées.

Le schéma relationnel de la relation **`Centres`** est le suivant :

**`Centres`**(id_centre: INT, nom_ville: TEXT, latitude: FLOAT, longitude: FLOAT, altitude: FLOAT)

Le schéma relationnel de la relation **`Mesures`** est le suivant :

**`Mesures`**(id_mesure: INT, id_centre: INT, date_mesure: DATE, temperature: FLOAT, pression: INT, pluviometrie: INT)

Les tables **`Centres`** et **`Mesures`** ont été créées par un fichier joint, et peuvent être directement consultées et modifiées. 

<div style="border:3px solid #058dad; border-radius:2pt; color:#404242; padding:16px; background-color: #edfafc">

#### Syntaxe SQL : Sélectionner des enregistrements
- Sélectionner l'ensemble des enregistrements d'une table :
```sql
SELECT * FROM Table;
```
- Sélectionner des enregistrements :
```sql
SELECT attributs FROM Table WHERE Selecteur;
```
- Sélectionner des enregistrements dans un ordre ascendant :
```sql
SELECT attributs FROM Table WHERE Selecteur ORDER BY attribut ASC;
```
- Sélectionner des enregistrements sans doublon :
```sql
SELECT DISTINCT attributs FROM Table WHERE Selecteur;
```
    
    
</div>

<div style="border-left:3px solid #058dad; border-radius:2pt; color:#404242; padding:16px; background-color: #f5dad0">
   
#### Question
1. Écrire une instruction permettant d'afficher tous les attributs et tous les enregistrements de la relation **Centres**
</div>

In [1]:
SELECT * FROM Centres;

id_centre	nom_ville	latitude	longitude	altitude
126	Le Puy-en-Velay	45.042	3.888	744
138	Grenoble	45.185	5.723	550
185	Tignes	45.469	6.909	2594
213	Amiens	49.894	2.293	60
263	Brest	48.388	-4.49	52
317	Gérardmer	48.073	6.879	855
459	Nice	43.706	7.262	260

<div style="border-left:3px solid #058dad; border-radius:2pt; color:#404242; padding:16px; background-color: #f5dad0">
   
#### Question
2. Écrire une instruction permettant d'afficher tous les attributs et tous les enregistrements de la relation **Mesures**
</div>

<div style="border-left:3px solid #058dad; border-radius:2pt; color:#404242; padding:16px; background-color: #f5dad0">
   
#### Question
3. Parmi les attributs de la relation **`Mesures`**, lequel peut être une clé primaire ? Justifier la réponse.
</div>


<div style="border:3px solid #058dad; border-radius:2pt; color:#404242; padding:16px; background-color: #edfafc">

#### Syntaxe SQL : Fonctions d'agrégation
    
- Valeur minimale d'un attribut :
```sql
SELECT MIN(attribut) FROM Table; 
``` 

- Valeur maximale d'un attribut :
```sql
SELECT MAX(attribut) FROM Table;
``` 

- Nombre d’enregistrements d'une table :
```sql
SELECT COUNT(*) FROM Table;
``` 

- Moyenne des valeurs d'un attribut :
```sql
SELECT AVG(attribut) FROM Table; 
``` 

- Somme des valeurs d'un attribut :
```sql
SELECT SUM(attribut) FROM Table; 
``` 
    
</div>

<div style="border-left:3px solid #058dad; border-radius:2pt; color:#404242; padding:16px; background-color: #f5dad0">
   
#### Question
4. Écrire une instruction permettant de dénombrer les enregistrements de la relation **Mesures**
</div>

<div style="border-left:3px solid #058dad; border-radius:2pt; color:#404242; padding:16px; background-color: #f5dad0">
   
#### Question
5. Écrire une requête SQL donnant la valeur moyenne des températures de la relation **Mesures**.
</div>

<div style="border-left:3px solid #058dad; border-radius:2pt; color:#404242; padding:16px; background-color: #f5dad0">
   
#### Question
6. Qu'affiche la requête suivante ?

```sql
SELECT * FROM Centres WHERE altitude > 500;
```


</div>

<div style="border-left:3px solid #058dad; border-radius:2pt; color:#404242; padding:16px; background-color: #f5dad0">
   
#### Question
7. On souhaite récupérer le nom de la ville des centres météorologiques situés à une altitude comprise entre 700 m et 1200 m.   
Écrire la requête SQL correspondante.

</div>

<div style="border-left:3px solid #058dad; border-radius:2pt; color:#404242; padding:16px; background-color: #f5dad0">
   
#### Question
8. Qu'affiche la requête suivante ?

```sql
SELECT * FROM Mesures WHERE date_mesure = "2021-10-30";
```
</div>

<div style="border-left:3px solid #058dad; border-radius:2pt; color:#404242; padding:16px; background-color: #f5dad0">
   
#### Question
9. On souhaite récupérer la liste des longitudes et des noms des villes des centres météorologiques dont la longitude est supérieure à 5. La liste devra être triée par ordre alphabétique des noms de ville.  
Écrire la requête SQL correspondante.

</div>

<div style="border:3px solid #058dad; border-radius:2pt; color:#404242; padding:16px; background-color: #edfafc">

#### Syntaxe SQL : Effectuer une jointure
- Effectuer une jointure :
```sql
SELECT attributs FROM TableA JOIN TableB ON TableA.cle1=TableB.cle2 WHERE Selecteur;
```
    
</div>

<div style="border-left:3px solid #058dad; border-radius:2pt; color:#404242; padding:16px; background-color: #f5dad0">
   
#### Question
10. Avec quels attributs peut-on faire une jointure entre la relation **`Centres`** et la relation **`Mesures`** ?
</div>


<div style="border-left:3px solid #058dad; border-radius:2pt; color:#404242; padding:16px; background-color: #f5dad0">
   
#### Question
11. Écrire une requête SQL donnant la liste des villes dans lesquelles on a enregistré une température inférieure à 10 °C. On utilisera le mot-clé `DISTINCT` afin d'éviter d'avoir des doublons. 
 

</div>

<div style="border-left:3px solid #058dad; border-radius:2pt; color:#404242; padding:16px; background-color: #f5dad0">
   
#### Question
12. Écrire une requête SQL donnant la ville dans laquelle on a enregistré la température minimale, ainsi que la valeur de cette température.
 

</div>

<div style="border:3px solid #058dad; border-radius:2pt; color:#404242; padding:16px; background-color: #edfafc">

#### Syntaxe SQL : Mettre à jour une table
- Insérer des enregistrements :
```sql
INSERT INTO Table (attribut1, attribut2) VALUES (valeur1 , valeur2);
```
- Modifier des enregistrements :
```sql 
UPDATE Table SET attribut1=valeur1, attribut2=valeur2 WHERE Selecteur;
```
- Supprimer des enregistrements :
```sql
DELETE FROM Table WHERE Selecteur;
```
    
</div>

<div style="border-left:3px solid #058dad; border-radius:2pt; color:#404242; padding:16px; background-color: #f5dad0">
   
#### Question
13. Écrire une requête SQL permettant d'ajouter une mesure prise le 8 novembre 2021 dans le centre numéro 138, où la température était de 11 °C, la pression de 1013 hPa et la pluviométrie de 0 mm. La donnée dont l'attribut est `id_mesure` aura pour valeur 3650.
 
</div>

<div style="border-left:3px solid #058dad; border-radius:2pt; color:#404242; padding:16px; background-color: #f5dad0">
   
#### Question
14. Vérifier que le nombre de mesures a bien augmenté par rapport au nombre initial de mesures.
 
</div>

<div style="border-left:3px solid #058dad; border-radius:2pt; color:#404242; padding:16px; background-color: #f5dad0">
   
#### Question
15. Que se passerait-il si l'on tentait d'exécuter l'instruction suivante :
   
```sql
DELETE FROM Centres WHERE id_centre=263;
```
 
</div>

<div style="border-left:3px solid #058dad; border-radius:2pt; color:#404242; padding:16px; background-color: #f5dad0">
   
#### Question
16. Expliquer ce que renvoie la requête SQL suivante :

```sql
SELECT * FROM Centres WHERE latitude = (SELECT MIN(latitude) FROM Centres);
```
 
</div>

<div style="border-left:3px solid #058dad; border-radius:2pt; color:#404242; padding:16px; background-color: #f5dad0">
   
#### Question
17. Écrire une requête SQL permettant d'afficher toutes les villes où la température a été supérieure au moins une fois à la température moyenne de la table **Mesures**.
 
</div>